File: lib/set-timestamp.py
"""Set timestamp and value fields in the FLOW state file.

Handles mid-phase timestamp fields and Code phase task updates.
Dot-notation paths navigate nested objects. Numeric path segments
index arrays. NOW is a magic value replaced with current Pacific Time timestamp.

Usage:
  bin/flow set-timestamp --set <path>=<value> [--set <path>=<value> ...]

Examples:
  bin/flow set-timestamp --set design.approved_at=NOW
  bin/flow set-timestamp --set plan.tasks.3.status=in_progress --set plan.tasks.3.started_at=NOW

Output (JSON to stdout):
  Success: {"status": "ok", "updates": [{"path": "...", "value": "..."}]}
  Error:   {"status": "error", "message": "..."}
"""

import argparse
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from flow_utils import current_branch, now, project_root


def _set_nested(obj, path_parts, value):
    """Navigate a nested dict/list by path parts and set the final value.

    Numeric path segments are treated as array indexes (0-based).
    """
    current = obj
    for part in path_parts[:-1]:
        if isinstance(current, list):
            try:
                index = int(part)
            except ValueError:
                raise KeyError(f"Expected numeric index for list, got '{part}'")
            if index < 0 or index >= len(current):
                raise IndexError(f"Index {index} out of range (list has {len(current)} items)")
            current = current[index]
        elif isinstance(current, dict):
            if part in current:
                current = current[part]
            else:
                raise KeyError(f"Key '{part}' not found")
        else:
            raise KeyError(f"Cannot navigate into {type(current).__name__} with key '{part}'")

    final_key = path_parts[-1]
    if isinstance(current, list):
        try:
            index = int(final_key)
        except ValueError:
            raise KeyError(f"Expected numeric index for list, got '{final_key}'")
        if index < 0 or index >= len(current):
            raise IndexError(f"Index {index} out of range (list has {len(current)} items)")
        current[index] = value
    elif isinstance(current, dict):
        current[final_key] = value
    else:
        raise KeyError(f"Cannot set key '{final_key}' on {type(current).__name__}")


def apply_updates(state, set_args):
    """Apply a list of path=value updates to the state dict.

    Returns (state, updates_list) where updates_list records what was set.
    """
    updates = []
    for assignment in set_args:
        if "=" not in assignment:
            raise ValueError(f"Invalid format '{assignment}' — expected path=value")

        path, value = assignment.split("=", 1)
        path_parts = path.split(".")

        if value == "NOW":
            value = now()
        elif value.isdigit():
            value = int(value)

        _set_nested(state, path_parts, value)
        updates.append({"path": path, "value": value})

    return state, updates


def main():
    parser = argparse.ArgumentParser(description="Set state file fields")
    parser.add_argument("--set", dest="set_args", action="append", required=True,
                        help="path=value (use NOW for current timestamp)")
    args = parser.parse_args()

    root = project_root()
    branch = current_branch()

    if not branch:
        print(json.dumps({
            "status": "error",
            "message": "Could not determine current branch",
        }))
        sys.exit(1)

    state_path = root / ".flow-states" / f"{branch}.json"

    if not state_path.exists():
        print(json.dumps({
            "status": "error",
            "message": f"No state file found: {state_path}",
        }))
        sys.exit(1)

    try:
        state = json.loads(state_path.read_text())
    except Exception as e:
        print(json.dumps({
            "status": "error",
            "message": f"Could not read state file: {e}",
        }))
        sys.exit(1)

    try:
        state, updates = apply_updates(state, args.set_args)
    except (KeyError, IndexError, ValueError) as e:
        print(json.dumps({
            "status": "error",
            "message": str(e),
        }))
        sys.exit(1)

    state_path.write_text(json.dumps(state, indent=2))
    print(json.dumps({"status": "ok", "updates": updates}))


if __name__ == "__main__":
    main()

File: src/commands/set_timestamp.rs
use serde_json::{json, Value};

use crate::flow_paths::FlowPaths;
use crate::git::resolve_branch;
use crate::lock::mutate_state;
use crate::utils::now;

/// A single path=value update that was applied.
#[derive(Debug)]
pub struct Update {
    pub path: String,
    pub value: Value,
}

/// Closed enumeration of the step-counter fields that
/// trigger an account-window step snapshot capture. The single
/// source of truth — code reads this when deciding whether to
/// append to `phases.<n>.step_snapshots[]` after a
/// `set-timestamp` call.
const STEP_COUNTER_FIELDS: &[&str] = &["code_task", "review_step", "learn_step", "complete_step"];

/// Returns `true` when `field` names one of the recognized step
/// counters. Whitespace and case are not normalized — callers pass
/// the field name straight from CLI argument parsing where
/// `--set <field>=<value>` already produced an exact slice of the
/// argument before the `=`.
pub fn is_step_counter_field(field: &str) -> bool {
    STEP_COUNTER_FIELDS.contains(&field)
}

/// Navigate a nested JSON Value by dot-path parts and set the final value.
///
/// Numeric path segments are treated as array indexes (0-based).
pub fn set_nested(obj: &mut Value, path_parts: &[&str], value: Value) -> Result<(), String> {
    if path_parts.is_empty() {
        return Err("Empty path".to_string());
    }

    let (intermediate, final_key) = path_parts.split_at(path_parts.len() - 1);

    let mut current = obj;
    for part in intermediate {
        current = match current {
            Value::Array(arr) => {
                let index: usize = match part.parse() {
                    Ok(n) => n,
                    Err(_) => {
                        return Err(format!("Expected numeric index for list, got '{}'", part))
                    }
                };
                if index >= arr.len() {
                    return Err(format!(
                        "Index {} out of range (list has {} items)",
                        index,
                        arr.len()
                    ));
                }
                &mut arr[index]
            }
            Value::Object(map) => map
                .get_mut(*part)
                .ok_or_else(|| format!("Key '{}' not found", part))?,
            Value::Null => {
                return Err(format!("Cannot navigate into NoneType with key '{}'", part))
            }
            Value::Bool(_) => return Err(format!("Cannot navigate into bool with key '{}'", part)),
            Value::Number(_) => {
                return Err(format!("Cannot navigate into int with key '{}'", part))
            }
            Value::String(_) => {
                return Err(format!("Cannot navigate into str with key '{}'", part))
            }
        };
    }

    let key = final_key[0];
    match current {
        Value::Array(arr) => {
            let index: usize = match key.parse() {
                Ok(n) => n,
                Err(_) => return Err(format!("Expected numeric index for list, got '{}'", key)),
            };
            if index >= arr.len() {
                return Err(format!(
                    "Index {} out of range (list has {} items)",
                    index,
                    arr.len()
                ));
            }
            arr[index] = value;
        }
        Value::Object(map) => {
            map.insert(key.to_string(), value);
        }
        Value::Null => return Err(format!("Cannot set key '{}' on NoneType", key)),
        Value::Bool(_) => return Err(format!("Cannot set key '{}' on bool", key)),
        Value::Number(_) => return Err(format!("Cannot set key '{}' on int", key)),
        Value::String(_) => return Err(format!("Cannot set key '{}' on str", key)),
    }

    Ok(())
}

/// Validate that code_task can only increment by 1 or reset to 0.
pub fn validate_code_task(state: &Value, new_value: i64) -> Result<(), String> {
    if new_value == 0 {
        return Ok(());
    }
    let current = state.get("code_task").and_then(|v| v.as_i64()).unwrap_or(0);
    if new_value != current.saturating_add(1) {
        let hint = if new_value == current.saturating_add(2) {
            format!(
                "--set code_task={} --set code_task={}",
                current.saturating_add(1),
                new_value
            )
        } else {
            format!(
                "--set code_task={} --set code_task={} ... --set code_task={}",
                current.saturating_add(1),
                current.saturating_add(2),
                new_value
            )
        };
        return Err(format!(
            "code_task can only increment by 1. Current: {}, attempted: {}. \
             Use multiple --set args in one call for atomic groups: {}",
            current, new_value, hint
        ));
    }
    Ok(())
}

/// Apply a list of path=value updates to the state Value.
///
/// Returns the list of updates that were applied.
pub fn apply_updates(state: &mut Value, set_args: &[String]) -> Result<Vec<Update>, String> {
    let mut updates = Vec::new();

    for assignment in set_args {
        let eq_pos = assignment
            .find('=')
            .ok_or_else(|| format!("Invalid format '{}' — expected path=value", assignment))?;

        let path = &assignment[..eq_pos];
        let raw_value = &assignment[eq_pos + 1..];

        let value: Value = if raw_value == "NOW" {
            Value::String(now())
        } else if let Ok(n) = raw_value.parse::<i64>() {
            json!(n)
        } else {
            Value::String(raw_value.to_string())
        };

        let path_parts: Vec<&str> = path.split('.').collect();

        if path_parts == ["code_task"] {
            let int_val = match value.as_i64() {
                Some(n) => n,
                None => return Err(format!("code_task must be an integer, got '{}'", raw_value)),
            };
            validate_code_task(state, int_val)?;
        }

        set_nested(state, &path_parts, value.clone())?;
        updates.push(Update {
            path: path.to_string(),
            value,
        });
    }

    Ok(updates)
}

/// Outcome of [`run_impl_main`]: a JSON payload (success or error
/// shape) and a paired exit code.
pub type RunOutcome = (Value, i32);

/// Testable core of the set-timestamp command. Returns the payload the
/// CLI wrapper would print plus the exit code. Tests pass tempdir
/// `root`/`cwd` to bypass cwd-scope drift and the on-disk state file.
///
/// Branch resolution uses [`resolve_branch`] (consults
/// `FLOW_SIMULATE_BRANCH` and falls back to `git branch --show-current`
/// in the process cwd). The `None` branch is covered by subprocess
/// tests that spawn the binary with `current_dir` pointing at a
/// non-git tempdir.
pub fn run_impl_main(
    set_args: &[String],
    branch_override: Option<&str>,
    root: &std::path::Path,
    cwd: &std::path::Path,
) -> RunOutcome {
    // Drift guard: set-timestamp is the general-purpose state mutator
    // for mid-phase fields. Writing to the state file from the wrong
    // subdirectory of a mono-repo would silently record values
    // against the wrong assumed scope. See
    // [`crate::cwd_scope::enforce`].
    if let Err(msg) = crate::cwd_scope::enforce(cwd, root) {
        return (json!({"status": "error", "message": msg}), 1);
    }

    let branch = match resolve_branch(branch_override, root) {
        Some(b) => b,
        None => {
            return (
                json!({
                    "status": "error",
                    "message": "Could not determine current branch"
                }),
                1,
            );
        }
    };

    // Per `.claude/rules/external-input-validation.md` "CLI subcommand
    // entry callsite discipline" + `.claude/rules/branch-path-safety.md`:
    // `--branch` is an external input. Slash-containing branches
    // (`feature/foo`, `dependabot/*`) and the empty string flow raw
    // from clap, so the panicking constructor would crash the CLI
    // with a backtrace. `try_new` returns `None` for invalid inputs;
    // translate that into a structured error.
    let state_path = match FlowPaths::try_new(root, &branch) {
        Some(p) => p.state_file(),
        None => {
            return (
                json!({
                    "status": "error",
                    "message": format!("Invalid branch name: {:?}", branch)
                }),
                1,
            );
        }
    };

    if !state_path.exists() {
        return (
            json!({
                "status": "error",
                "message": format!("No state file found: {}", state_path.display())
            }),
            1,
        );
    }

    let mut collected_updates: Vec<Update> = Vec::new();
    let mut apply_error: Option<String> = None;

    // Snapshot state before applying updates so a mid-way failure can
    // restore the original — `apply_updates` mutates in place.
    let home = crate::session_metrics::home_dir_or_empty();
    let result = mutate_state(&state_path, &mut |state| {
        let backup = state.clone();
        match apply_updates(state, set_args) {
            Ok(updates) => {
                // Per-step-counter snapshot: for each successful update
                // whose field is one of the five step counters, capture
                // a window snapshot and append it to
                // phases.<current_phase>.step_snapshots[].
                let current_phase = state
                    .get("current_phase")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string();
                if !current_phase.is_empty() {
                    for update in &updates {
                        if !is_step_counter_field(&update.path) {
                            continue;
                        }
                        let step = update.value.as_i64().unwrap_or(0);
                        let snap =
                            crate::per_flow_capture::capture_for_active_state(&home, state, root);
                        crate::session_metrics::append_step_snapshot(
                            state,
                            &current_phase,
                            step,
                            &update.path,
                            snap,
                        );
                    }
                }
                collected_updates = updates;
            }
            Err(e) => {
                *state = backup;
                apply_error = Some(e);
            }
        }
    });

    if let Some(msg) = apply_error {
        return (json!({"status": "error", "message": msg}), 1);
    }

    match result {
        Ok(_) => {
            let updates_json: Vec<Value> = collected_updates
                .iter()
                .map(|u| json!({"path": u.path, "value": u.value}))
                .collect();
            (json!({"status": "ok", "updates": updates_json}), 0)
        }
        Err(e) => {
            let msg = e.to_string();
            let message = if msg.contains("Invalid JSON") || msg.contains("JSON error") {
                format!("Could not read state file: {}", msg)
            } else {
                msg
            };
            (json!({"status": "error", "message": message}), 1)
        }
    }
}
